src/V3Fork.cpp

```diff
@@ -34,12 +46,351 @@
+
+        // Find node after last variable declaration
+        AstNode* initp = stmtp;
+        while (initp && VN_IS(initp, Var)) initp = initp->nextp();
+        UASSERT(stmtp, "Procedure lacks body");
+        UASSERT(initp, "Procedure lacks statements besides declarations");
+
+        AstNew* newp = new AstNew{m_procp->fileline(), nullptr};
+        newp->taskp(VN_AS(memberMap.findMember(m_instance.classp, "new"), NodeFTask));
+        newp->dtypep(m_instance.refDTypep);
+        newp->classOrPackagep(m_instance.classp);
+
+        AstNode* asgnp = new AstAssign{
+            m_procp->fileline(),
+            new AstVarRef{m_procp->fileline(), m_instance.handlep, VAccess::WRITE}, newp};
+
+        AstNode* inits = nullptr;  // Arguments need to be copied
+        for (AstVar* varp : m_captures) {
+            if (VL_LIKELY(varp->direction() != VDirection::INPUT)) continue;
+
+            AstMemberSel* const membersel = new AstMemberSel{
+                varp->fileline(),
+                new AstVarRef{varp->fileline(), m_instance.handlep, VAccess::WRITE},
+                varp->dtypep()};
+            membersel->name(varp->name());
+            membersel->varp(VN_AS(memberMap.findMember(m_instance.classp, varp->name()), Var));
+            AstNode* initAsgnp = new AstAssign{
+                varp->fileline(), membersel, new AstVarRef{varp->fileline(), varp, VAccess::READ}};
+            inits = AstNode::addNext(inits, initAsgnp);
+        }
+        if (inits) AstNode::addNext(asgnp, inits);
+
+        if (initp != stmtp) {
+            initp->addHereThisAsNext(asgnp);
+        } else {
+            AstNode::addNext(asgnp, static_cast<AstNode*>(initp));
+            stmtp = asgnp;
+        }
+
+        AstNode::addNext(static_cast<AstNode*>(m_instance.handlep), stmtp);
+        stmtpHandle.relink(m_instance.handlep);
+        m_modp->addStmtsp(m_instance.classp);
+    }
+
+    bool linked() const { return m_instance && m_instance.handlep->backp(); }
+
+private:
+    static string generateDynScopeClassName(const AstNode* fromp) {
+        // TODO: Ensure no collisions occur
+        string n = "__VDynScope__" + (!fromp->name().empty() ? (fromp->name() + "__") : "ANON__")
+                   + cvtToHex(fromp);
+        return n;
+    }
+
+    static string generateDynScopeHandleName(const AstNode* fromp) {
+        return "__VDynScope_" + (fromp->name().empty() ? cvtToHex(fromp) : fromp->name());
+    }
+
+    AstNode* getProcStmts() {
+        AstNode* stmtsp = nullptr;
+        if (!m_procp) return nullptr;
+        if (AstBegin* beginp = VN_CAST(m_procp, Begin))
+            stmtsp = beginp->stmtsp();
+        else if (AstNodeFTask* taskp = VN_CAST(m_procp, NodeFTask))
+            stmtsp = taskp->stmtsp();
+        else
+            v3fatal("m_procp is not a begin block or a procedure");
+        return stmtsp;
+    }
+
+    AstNodeModule* m_modp;
+    AstNode* m_procp;
+    std::set<AstVar*> m_captures;
+    DynScopeInstance m_instance;
```

Comment:
Move to top of class under // MEMBER comment, and document each.  Check all are inited. Perhaps some can be const?

Reply:
I thought that mostly applied to visitors but it definitely won't hurt to document some other classes. I moved those to the top, and decorated them with some comments.